cpp17/diff_match_patch_utils.cpp

```diff
@@ -0,0 +1,168 @@
+/*
+ * Diff Match and Patch
+ * Copyright 2018 The diff-match-patch Authors.
+ * https://github.com/google/diff-match-patch
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "diff_match_patch_utils.h"
+
+#include <codecvt>
+#include <locale>
+
+namespace NUtils {
+std::wstring safeMid(const std::wstring &str, std::size_t pos) {
+  return safeMid(str, pos, std::string::npos);
+}
+
+std::wstring safeMid(const std::wstring &str, std::size_t pos,
+                     std::size_t len) {
+  return (pos == str.length()) ? std::wstring() : str.substr(pos, len);
+}
+
+void replace(std::wstring &inString, const std::wstring &from,
+             const std::wstring &to) {
+  std::size_t pos = inString.find(from);
+  while (pos != std::wstring::npos) {
+    inString.replace(pos, from.length(), to);
+    pos = inString.find(from, pos + to.length());
+  }
+}
+
+wchar_t toHexUpper(wchar_t value) { return L"0123456789ABCDEF"[value & 0xF]; }
+
+std::wstring toPercentEncoding(wchar_t c, const std::wstring &exclude,
+                               const std::wstring &include) {
+  std::wstring retVal;
+
+  if (((c >= 0x61 && c <= 0x7A)     // ALPHA
+       || (c >= 0x41 && c <= 0x5A)  // ALPHA
+       || (c >= 0x30 && c <= 0x39)  // DIGIT
+       || c == 0x2D                 // -
+       || c == 0x2E                 // .
+       || c == 0x5F                 // _
+       || c == 0x7E                 // ~
+       || (exclude.find(c) != std::string::npos)) &&
+      (include.find(c) == std::string::npos)) {
+    retVal = std::wstring(1, c);
+  } else {
+    retVal = L'%';
+    retVal += toHexUpper((c & 0xf0) >> 4);
+    retVal += toHexUpper(c & 0xf);
+  }
+  return retVal;
+}
+
+std::wstring toPercentEncoding(
+    const std::wstring &input, const std::wstring &exclude /*= std::wstring()*/,
+    const std::wstring &include /*= std::wstring() */) {
+  if (input.empty()) return {};
+  std::wstring retVal;
+  retVal.reserve(input.length() * 3);
+
+  static_assert(sizeof(wchar_t) <= 4, "wchar_t is greater that 32 bit");
+
+  std::wstring_convert<std::codecvt_utf8<wchar_t> > utf8_conv;
```

Comment:
'std::codecvt_utf8<wchar_t,1114111,(std::codecvt_mode)0>': warning STL4017: std::wbuffer_convert, std::wstring_convert, and the <codecvt> header (containing std::codecvt_mode, std::codecvt_utf8, std::codecvt_utf16, and std::codecvt_utf8_utf16) are deprecated in C++17. (The std::codecvt class template is NOT deprecated.) The C++ Standard doesn't provide equivalent non-deprecated functionality; consider using MultiByteToWideChar() and WideCharToMultiByte() from <Windows.h> instead. You can define _SILENCE_CXX17_CODECVT_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.

Same in line 120